mmctools/wrf/preprocessing.py

```diff
@@ -340,3 +398,1067 @@ def download(self,datetimes,path=None,bounds={}):
+        missing_options = False
+        setup_keys = list(self.setup_dict.keys())
+        for key in required_fields:
+            if key not in setup_keys:
+                missing_keys.append(key)
+                missing_options = True
+                
+        if not missing_options:
+            if 'usgs' in self.setup_dict['geogrid_args']:
+                land_cat = 24
+            else:
+                land_cat = 21
+            namelist_defaults = {
+                       'geogrid_args' : '30s',
+                   'history_interval' : [60,60],
+                   'interval_seconds' : self.icbc_dict['interval_seconds'],
+                 'num_metgrid_levels' : self.icbc_dict['met_levels'],
+            'num_metgrid_soil_levels' : self.icbc_dict['soil_levels'],
+                    'input_from_file' : '.true.',
+                   'restart_interval' : 360,            
+                    'frames_per_file' : 1,            
+                              'debug' : 0,            
+                       'ts_locations' : 20,            
+                          'ts_levels' : self.setup_dict['num_eta_levels'],            
+                         'mp_physics' : 10,            
+                              'ra_lw' : 4,            
+                              'ra_sw' : 4,
+                               'radt' : int(self.setup_dict['dxy']/1000.0),
+                  'sf_sfclay_physics' : 1,
+                 'sf_surface_physics' : 2,
+                     'bl_pbl_physics' : 1,
+                         'cu_physics' : 1,
+                             'isfflx' : 1,
+                             'ifsnow' : 0,
+                             'icloud' : 0,
+               'surface_input_source' : 1,
+                    'num_soil_layers' : self.icbc_dict['soil_levels'],
+                       'num_land_cat' : land_cat,
+                   'sf_urban_physics' : 0,
+                          'w_damping' : 1,
+                           'diff_opt' : 1,
+                             'km_opt' : 4,
+                       'diff_6th_opt' : 2,
+                    'diff_6th_factor' : 0.12,
+                          'base_temp' : 290.0,
+                           'damp_opt' : 3,
+                              'zdamp' : 5000.0,
+                           'dampcoef' : 0.2,
+                              'khdif' : 0,
+                              'kvdif' : 0,
+                    'non_hydrostatic' : '.true.',
+                      'moist_adv_opt' : 1,
+                     'scalar_adv_opt' : 1,
+                        'tke_adv_opt' : 1,
+                    'h_mom_adv_order' : 5,
+                    'v_mom_adv_order' : 3,
+                    'h_sca_adv_order' : 5,
+                    'v_sca_adv_order' : 3,
+                     'spec_bdy_width' : 5,
+                          'spec_zone' : 1,
+                         'relax_zone' : 4,
+                'nio_tasks_per_group' : 0,
+                         'nio_groups' : 1,
+            }
```

Comment:
This is good for now, but I would like to make this even more human-readable in the future.

Reply:
Note to self: might want to create a namelist class that can 1) read in an existing namelist to re-use inputs; 2) map namelist integers and more meaningful model names; 3) put the namelist writer in there.

Reply:
Developing something like that wouldn't be too difficult and would be awesome to have!